Adds documentation and sample app for msal go (#22)
Azure-Samples/msal-managed-identity: src/go/readme.md

---
services: active-directory
platforms: go
author: andrewohart
level: 400
client: Azure functions
service: Azure Key Vault
endpoint: Microsoft identity platform
page_type: sample
languages:
  - go  
products:
  - azure
  - azure-active-directory  
  - go
  - managed identity
  - key vault
description: "This file is to give a general overview of Managed Identity for MSAL GO and the various sources and samples that can be used"
---
# GO - Samples Overview

This document is to give samples and step by step instructions on how to run the various Managed Identity Source samples

## Overview

MSAL GO supports acquiring tokens through the managed identity capability when used with applications running inside Azure infrastructure.
Samples can all be tested by downloading the [MSAL (Microsoft Authentication Library) for GO](https://github.com/AzureAD/microsoft-authentication-library-for-go)

The Sources with samples are as follows:

> - IMDS

Reviewer: App service is missing here

  miSystemAssigned, err := mi.New(mi.SystemAssigned())
  if err != nil {
      log.Fatal(err)
  }
```

You can then call the `AcquireToken` function to get a token or throw an error

```go
  // Get the access token using MSAL, or an error if there was one, you can see this in the sample app
  accessToken, err := miClient.AcquireToken(context.Background(), "https://vault.azure.net")
  if err != nil {
      log.Fatalf("failed to acquire token: %v", err)
      return
  }
```

The `AcquireToken` function in the `managedidentity.go` class demonstrates how to take advantage of the Managed Identity Client for calling Microsoft Key Vault without having to worry about secrets or certificates.

```go
  // In AcquireToken, when claims are empty, we get token from the cache, otherwise acquire a new one
  if o.claims == "" {
      storageTokenResponse, err := cacheManager.Read(ctx, c.authParams)
      if err != nil {
          return base.AuthResult{}, err
      }
      ar, err := base.AuthResultFromStorage(storageTokenResponse)
      if err == nil {
          ar.AccessToken, err = c.authParams.AuthnScheme.FormatAccessToken(ar.AccessToken)
          return ar, err
      }
  }
```

> Then depending on the Source, the Request will be created

Reviewer: ```suggestion
Then depending on the Source, the Request will be created.
```

Creation of the Request is different per Source, but usually consists of getting the URL from environment variables and adding

> in the correct Headers

Reviewer: This should be a single line.

> From here we attempt to get the token for the request

Reviewer: ```suggestion
From here MSAL attempt to get the token for the request.
```

```go
  tokenResponse, err := c.getTokenForRequest(req, resource)
  if err != nil {
      return base.AuthResult{}, err
  }
```

> We then save to the cache and return the Auth Result

Reviewer: ```suggestion
MSAL will then save to the cache and return the Auth Result.
```

> From here we print the Auth Result

Reviewer: Is this done by MSAL? Or is this a suggestion to the user to print?

Author: This is about how a user can do it, i will clarify

---
services: active-directory
platforms: go
author: andrewohart
level: 400
client: Azure functions
service: Azure Key Vault
endpoint: Microsoft identity platform
page_type: sample
languages:
  - go  
products:
  - azure
  - azure-active-directory  
  - go
  - managed identity
  - key vault
description: "This sample showcases how to develop an Azure function that gets a secret from a key Vault using Managed Identities."
---
# GO - Acquire a secret from an Azure Key Vault using Azure Managed Identity

## About this sample

### Overview

> This sample showcases how to acquire a secret from an Azure Key Vault using Azure Managed Identity. It shows you how to use the [Azure Functions](https://learn.microsoft.com/azure/azure-functions/functions-overview?pivots=programming-language-java) and acquire a token for an Azure Key Vault resource.

Reviewer: The link here is for java. Is that supposed to be the java link?

Author: Nope will update!

### Step 2:  Modify the Key Vault URI and Secret name values in the code

Following are the changes you need to make:

- In the [`handler.go`](https://github.com/Azure-Samples/msal-managed-identity/blob/main/src/go/sample/AcquireTokenMSI.go) file under the getSecretFromAzureVault method modify the following values, 

    ```go
        keyVaultUri := "your-key-vault-uri"
        secretName := "your-secret-name"
    ```

- Change these to match your key vault uri and secret name. These can be found in the following locations:

1. Key Vault URI - In your Azure home page, go to your key vault, on the Overview page our key vault URI can be found under **'Essentials'**
1. Secret Name - On the Key Vault Overview page, go to the Panel on the left and expand the **'Objects'** dropdown  
Click into **'Secrets'**  
Click into the secret you want to use  
Click ont the version you would like to use  
Copy the part after the key vault URI and use that as your secret name  

> ## Publish your function

Reviewer: Since the app service is not implemented, will it be relevant to have a sample for the functions? Can we have a sample for VM instead since that is implemented?

Author: I marked this as a draft PR so its ready to go when App Service is implemented. The other docs in this sample project are using AppService so I think we want to keep this the same, and then the other PR that is open is more about using the library
